Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81113
* __->__ #78616
* #78615
* #78350
* #80219

This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

lezcano committed:
Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals

This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

Lezcano commented on torch/_prims/__init__.py:
```diff
@@ -2467,6 +2473,73 @@ def _full_like_aten(
+    )
+
+    shape_Vh = batch + (n if full_matrices else k, n)
+    # The CPU backend returns V, but the cuSolver backend returns V^H
+    # TODO The MAGMA backend returns V, so this is wrong if used with the MAGMA backend
+    is_cuda = (A.device.type == "cuda")
+    strides_Vh = contig_strides(shape_Vh, row_major=is_cuda)
+    Vh = TensorMeta(shape=shape_Vh, strides=strides_Vh, dtype=A.dtype, device=A.device)
```
Not clear what to do here.

Lezcano commented on torch/_prims/__init__.py:
```diff
@@ -2467,6 +2473,73 @@ def _full_like_aten(
+    return U, S, Vh
+
+
+def _svd_aten(
+    A: TensorLikeType, *, full_matrices: bool
+) -> Tuple[Tensor, Tensor, Tensor]:
```
Not sure whether we should model here the `compute_uv` parameter. On the one hand, it is useful when computing `svdvals`, but only when no gradients are required. As such, this optimisation is not fully contemplated at the moment in PrimTorch, given that we are not considering gradients yet.
Should we add the `compute_uv` parameter here (as we do in `_linalg_svd` in PyTorch really)?

mruberry replied:
Good thing we can always change our mind later -- let's try it without `compute_uv` for now

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -2416,6 +2422,73 @@ def _full_like_aten(
+_svd_doc = """
+    Returns the SVD of a matrix or batch of matrices.
+
+    The `full_matrices` flag controls whether the full or reduced SVD decomposition is returned.
```
Future question: should the full and reduced SVD be different prims?

Lezcano replied:
Probably not. If we're going in the direction of ending up exposing something that smells like the LAPCK / cuSolver API and then write the dispatch in Python, we probably want them to be the same operation

mruberry reviewed (approved):
Cool! How, exactly, we model these prims is an interesting discussion for the future

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g

lezcano committed:
Update on "Add prim.svd, refs.linalg.svd, and refs.linalg.svdvals"


This is the first prim / ref added that has multiple returns.
There is an issue with `out_wrapper_multi` as currently implemented
(left a note). It assumes that the API is `svd(X, U=U, S=S, Vh=Vh)`,
when it's actually `svd(X, out=(U, S, Vh))`.

Even more, if we want to model PyTorch exactly, it should return a
`torch.return_types.svd`, rather than a `Tuple`.

There is an issue with
As per title

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g
